[inductor] [cpp] Support vectorization for score and mask in FlexAttention CPU

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143638

## Description
We generate vectorized kernel for score and mask in FlexAttention with this PR.

## Modification
The main change include:
- For the input and output buffer to the mask and score function, instead of passing scalars, we pass tensors to it.
- For the mask function, the original function which works on a scalar only includes the logic of calculating the mask value. The PR added the logic of applying the mark to the qk_data tensor into the graph and then leverage the CPP backend to generate vectorized kernels.
  The original mask graph:
  ```python
  def mask_fn(b, h, q_idx, kv_idx):
      mask = q_idx >= kv_idx
      return mask
  ```
  The converted_mask_graph should be:
  ```python
  def converted_mask_fn(qk_data, b, h, q_idx, kv_idx):
      mask = q_idx >= kv_idx
      qk_data = torch.where(mask, qk_data, torch.full_like(qk_data, -float("inf")))
      return qk_data
  ```

## Benchmark
For q, k, v of shape: `[1, 32, 1024, 128]`, using 40 CPU cores, we observe over 20x speedup compared with the non vectorized version for both `is_causal` = `False` and `True`.


## Test plan
The existing FlexAttention UTs (`test/inductor/test_flex_attention.py`, `test/inductor/test_flex_decoding.py`) can cover the change in this PR.

## Output code

**Code before this PR is in scalar version:**

```cpp
// apply score mod function
for (int64_t row = 0; row < cur_qSplitSize; ++row) {
    for (int64_t col = 0; col < cur_kvSplitSize; col++) {
    std::vector<int64_t> b_idx = {i};
    std::vector<int64_t> h_idx = {j};
    std::vector<int64_t> q_idx = {m+row};
    int64_t phisical_kv_idx = n+col;
    if (use_kv_indice) {
        phisical_kv_idx= *kv_logical_data * kvBlockSize + col;
    }
    std::vector<int64_t> kv_idx = {phisical_kv_idx};
    accum_t* in_ptr0 = qk_data + row * cur_kvSplitSize + col;
    auto in_ptr1 = b_idx.data();
    auto in_ptr2 = h_idx.data();
    auto in_ptr3 = q_idx.data();
    auto in_ptr4 = kv_idx.data();

    accum_t* out_ptr0 = in_ptr0;
    {
        {
            {
                auto tmp0 = in_ptr0[static_cast<int64_t>(0L)];
                out_ptr0[static_cast<int64_t>(0L)] = tmp0;
            }
        }
    }

    }
}
// Apply block mask, fill unused with -inf
for (int64_t row = 0; row < cur_qSplitSize; ++row) {
    for (int64_t col = 0; col < cur_kvSplitSize; col++) {
    std::vector<int64_t> b_idx = {i};
    std::vector<int64_t> h_idx = {j};
    std::vector<int64_t> q_idx = {m+row};
    int64_t phisical_kv_idx = n+col;
    if (use_kv_indice) {
        phisical_kv_idx= *kv_logical_data * kvBlockSize + col;
    }
    std::vector<int64_t> kv_idx = {phisical_kv_idx};
    accum_t* qk_block = qk_data + row * cur_kvSplitSize + col;
    auto in_ptr1 = b_idx.data();
    auto in_ptr2 = h_idx.data();
    auto in_ptr3 = q_idx.data();
    auto in_ptr4 = kv_idx.data();

    std::vector<int64_t> temp = {0};
    int64_t* out_ptr1 = temp.data();
    {
        {
            {
                auto tmp0 = static_cast<bool>(true);
                out_ptr1[static_cast<int64_t>(0L)] = tmp0;
            }
        }
    }

    *qk_block = *out_ptr1 != 0
                    ? *qk_block
                    : -std::numeric_limits<accum_t>::infinity();
    }
}
```

**Code after this PR will be vectorized:**
```cpp
accum_t* in_ptr0 = qk_data;

auto in_ptr1 = b_idx.data();
auto in_ptr2 = h_idx.data();
auto in_ptr3 = q_idx.data();
auto in_ptr4 = kv_idx.data();

// apply score mod function
{

    accum_t* out_ptr0 = in_ptr0;
    {
        #pragma GCC ivdep
        for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(cur_qSplitSize); x0+=static_cast<int64_t>(1L))
        {
            for(int64_t x1=static_cast<int64_t>(0L); x1<static_cast<int64_t>(cur_kvSplitSize); x1+=static_cast<int64_t>(16L))
            {
                {
                    if(C10_LIKELY(x1 >= static_cast<int64_t>(0) && x1 < static_cast<int64_t>(16L*(c10::div_floor_integer(static_cast<int64_t>(cur_kvSplitSize), static_cast<int64_t>(16L))))))
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<int64_t>(x1 + cur_kvSplitSize*x0), static_cast<int64_t>(16));
                        tmp0.store(out_ptr0 + static_cast<int64_t>(x1 + cur_kvSplitSize*x0));
                    }
                    if(C10_UNLIKELY(x1 >= static_cast<int64_t>(16L*(c10::div_floor_integer(static_cast<int64_t>(cur_kvSplitSize), static_cast<int64_t>(16L)))) && x1 < static_cast<int64_t>(cur_kvSplitSize)))
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<int64_t>(x1 + cur_kvSplitSize*x0), static_cast<int64_t>(cur_kvSplitSize + ((-16L)*(c10::div_floor_integer(static_cast<int64_t>(cur_kvSplitSize), static_cast<int64_t>(16L))))));
                        tmp0.store(out_ptr0 + static_cast<int64_t>(x1 + cur_kvSplitSize*x0), static_cast<int64_t>(cur_kvSplitSize + ((-16L)*(c10::div_floor_integer(static_cast<int64_t>(cur_kvSplitSize), static_cast<int64_t>(16L))))));
                    }
                }
            }
        }
    }

}

// Apply block mask, fill unused with -inf
{

    accum_t* out_ptr1 = in_ptr0;
    {
        #pragma GCC ivdep
        for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(cur_qSplitSize); x0+=static_cast<int64_t>(1L))
        {
            for(int64_t x1=static_cast<int64_t>(0L); x1<static_cast<int64_t>(cur_kvSplitSize); x1+=static_cast<int64_t>(16L))
            {
                {
                    if(C10_LIKELY(x1 >= static_cast<int64_t>(0) && x1 < static_cast<int64_t>(16L*(c10::div_floor_integer(static_cast<int64_t>(cur_kvSplitSize), static_cast<int64_t>(16L))))))
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<int64_t>(x1 + cur_kvSplitSize*x0), static_cast<int64_t>(16));
                        auto tmp1 = static_cast<bool>(true);
                        auto tmp2 = -std::numeric_limits<float>::infinity();
                        auto tmp3 = at::vec::VecMask<float,1>::from(tmp1);
                        auto tmp4 = at::vec::Vectorized<float>(tmp2);
                        auto tmp5 = decltype(tmp0)::blendv(tmp4, tmp0, tmp3.template cast<float,1>());
                        tmp5.store(out_ptr1 + static_cast<int64_t>(x1 + cur_kvSplitSize*x0));
                    }
                    if(C10_UNLIKELY(x1 >= static_cast<int64_t>(16L*(c10::div_floor_integer(static_cast<int64_t>(cur_kvSplitSize), static_cast<int64_t>(16L)))) && x1 < static_cast<int64_t>(cur_kvSplitSize)))
                    {
                        for (int64_t x1_tail = static_cast<int64_t>(16L*(c10::div_floor_integer(static_cast<int64_t>(cur_kvSplitSize), static_cast<int64_t>(16L))));x1_tail < static_cast<int64_t>(cur_kvSplitSize); x1_tail++)
                        {
                            auto tmp0 = in_ptr0[static_cast<int64_t>(x1_tail + cur_kvSplitSize*x0)];
                            auto tmp1 = static_cast<bool>(true);
                            auto tmp2 = -std::numeric_limits<float>::infinity();
                            auto tmp3 = tmp1 ? tmp0 : tmp2;
                            out_ptr1[static_cast<int64_t>(x1_tail + cur_kvSplitSize*x0)] = tmp3;
                        }
                    }
                }
            }
        }
    }

}
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w committed:
Update

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -871,14 +890,36 @@ def lower_cpu(
+    # Mark symbols > 1 to ensure broadcasting is always applied.
+    # This avoids treating them as equal when `eq(var, 1)` is evaluated in `broadcast_symbolic_shapes`.
+    shape_env.var_to_range[cur_qSplitSize] = ValueRanges(2, int_oo)
+    shape_env.var_to_range[cur_kvSplitSize] = ValueRanges(2, int_oo)
```
Is it possible `cur_qSplitSize` decided as 1 during runtime?

chunyuan-w replied:
In the generated code, we will always broadcast to the shape of the score tensor which is `[cur_qSplitSize, cur_kvSplitSize]`. The code will still be correct if `cur_qSplitSize` or `cur_kvSplitSize` is 1.

leslie-fang-intel commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -892,16 +933,74 @@ def lower_cpu(
-            ("kv_idx", torch.int64),
+        create_placeholder(name, dtype, query.get_device(), size)
+        for name, dtype, size in [
+            ("score", score_dtype, [cur_qSplitSize, cur_kvSplitSize]),
```
why `score` needed in `mask_graph_placeholder_inps`?

chunyuan-w replied:
The `mask_graph` only describes the below function:
```
def mask_fn(b, h, q_idx, kv_idx):
    mask = q_idx >= kv_idx
    return mask
```

What we need to generate the code for is the below function:
```
def converted_mask_fn(qk_data, b, h, q_idx, kv_idx):
    mask = q_idx >= kv_idx
    score= torch.where(mask, score, torch.full_like(score, -float("inf")))
    return score
```
We add `score` to generate the code for `score= torch.where(mask, score, torch.full_like(score, -float("inf")))`

leslie-fang-intel commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -892,16 +933,74 @@ def lower_cpu(
+        converted = torch.fx.GraphModule(mask_graph.graph_module, graph)
+        return converted
+
+    converted_mask_graph_module = convert_mask_graph_module(mask_graph)
```
cc @drisspg, do you think it's ok for us to modify the `mask_graph` or should we copy a new graph to do the modification here?

drisspg replied:
The mask if used both in forward and above, I think I commented above that its confusing to me to modify `create_placeholders`  to actually take in size.

I suppose that is valid since you are also modifying the graph. I think a separate graph that is only used for codegen w/ the same semantics of the captured graph makes more sense

chunyuan-w replied:
I've updated the code to create a deepcopy of the `mask_graph.graph_module` and modify this deepcopied one. Could you please re-review if this is okay now?

As for the `size` used in `create_placeholders`, here're more explanations: https://github.com/pytorch/pytorch/pull/143638#discussion_r1924597873

jgong5 commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -892,16 +933,74 @@ def lower_cpu(
+    #       mask = q_idx >= kv_idx
+    #       qk_data = torch.where(mask, qk_data, torch.full_like(qk_data, -float("inf")))
+    #       return qk_data
+    def convert_mask_graph_module(mask_graph):
```
The mask function (and the corresponding traced graph) is provided by the user. Is this function able to tensorize any mask function provided by the users?

chunyuan-w replied:
Yes, for the mask itself, the function supports any mask function defined by the user. As for the way to apply the calculated mask on the `qk_data` (or the `score` tensor), it's always the same no matter what the mask function is.

drisspg replied:
I am still pretty confused as to why you are inserting the score arg and creating this new fx graph as opposed to updating the template to handle this behavior similiar to hows its done for the triton template

drisspg replied:
Yeah alright you have convinced me

chunyuan-w commented:
Hi @drisspg could you please review this PR when you get time?

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -83,10 +85,21 @@ def flex_attention_grid(batch_size, q_heads, num_queries, d_model, meta):
+        layout=FixedLayout(
+            device,
+            dtype,
+            size if size else [],
```
Since we create the subgraphs using scalars we dont any real size or stride information, why this change?

chunyuan-w replied:
Before this PR, we create the subgraphs using scalars. In this PR, in order to generate vectorized code, we create subgraphs using tensors and thus we changed this function
https://github.com/pytorch/pytorch/blob/gh/chunyuan-w/3/head/torch/_inductor/kernel/flex_attention.py#L918-L922

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -871,14 +890,36 @@ def lower_cpu(
 
     fake_buffers: List[Buffer] = []  # noqa: F821
+
+    # The value of cur_qSplitSize and cur_kvSplitSize are decided during runtime.
```
Maybe ahve one named '[Note]' and point this comment there

chunyuan-w replied:
I've added a `[Note] Handle the case where the split sizes are not statically known` comment and inside `CppFlexAttentionTemplate` where we replace the string, I pointed this comment to make it clearer.

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w commented:
Hi @drisspg thanks for your reply. I have addressed all the comments and could you please re-review it to see if it's good to land or any other changes needed?

drisspg reviewed (approved):
Sorry I went on PTO for a few days and thought I had approved, thanks for the ping

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w committed:
Update

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge
